File: examples/hello-world/gabriel-greeting-app-kotlin-compose/app/src/main/kotlin/org/organicprogramming/gabriel/greeting/kotlincompose/controller/GreetingController.kt
package org.organicprogramming.gabriel.greeting.kotlincompose.controller

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import org.organicprogramming.gabriel.greeting.kotlincompose.model.AppPlatformCapabilities
import org.organicprogramming.gabriel.greeting.kotlincompose.model.GabrielHolonIdentity
import org.organicprogramming.gabriel.greeting.kotlincompose.model.GreetingUiState
import org.organicprogramming.gabriel.greeting.kotlincompose.model.LanguageOption
import org.organicprogramming.gabriel.greeting.kotlincompose.model.normalizedTransportSelection
import org.organicprogramming.gabriel.greeting.kotlincompose.model.preferredHolon
import org.organicprogramming.gabriel.greeting.kotlincompose.runtime.GreetingHolonConnection
import org.organicprogramming.gabriel.greeting.kotlincompose.runtime.HolonCatalog
import org.organicprogramming.gabriel.greeting.kotlincompose.runtime.HolonConnector
import java.util.concurrent.atomic.AtomicBoolean

class GreetingController(
    private val catalog: HolonCatalog,
    private val connector: HolonConnector,
    private val capabilities: AppPlatformCapabilities = AppPlatformCapabilities.desktopCurrent(),
    initialTransport: String? = System.getenv("OP_ASSEMBLY_TRANSPORT"),
    dispatcher: CoroutineDispatcher = Dispatchers.IO,
) {
    private val scope = CoroutineScope(SupervisorJob() + dispatcher)
    private val initialized = AtomicBoolean(false)
    private val _state = MutableStateFlow(
        GreetingUiState(
            transport = normalizedTransportSelection(initialTransport),
        ),
    )

    val state: StateFlow<GreetingUiState> = _state.asStateFlow()

    private var connection: GreetingHolonConnection? = null
    private var startFuture: kotlinx.coroutines.Deferred<Unit>? = null
    private var disposed = false
    private var connectionGeneration = 0
    private var loadGeneration = 0
    private var greetGeneration = 0

    suspend fun initialize() {
        if (!initialized.compareAndSet(false, true)) {
            return
        }
        refreshHolons()
        loadLanguages(greetAfterLoad = false)
        if (state.value.selectedLanguageCode.isNotBlank()) {
            greet()
        }
    }

    suspend fun refreshHolons() {
        val previousSelection = state.value.selectedHolon?.slug
        try {
            val discovered = catalog.discover()
            val selected = when {
                discovered.isEmpty() -> null
                previousSelection != null -> discovered.firstOrNull { it.slug == previousSelection }
                else -> null
            } ?: preferredHolon(discovered) ?: discovered.firstOrNull()

            _state.update {
                it.copy(
                    availableHolons = discovered,
                    selectedHolon = selected,
                    connectionError = null,
                )
            }
        } catch (error: Throwable) {
            _state.update {
                it.copy(
                    availableHolons = emptyList(),
                    selectedHolon = null,
                    connectionError = "Failed to discover Gabriel holons: $error",
                )
            }
        }
    }

    suspend fun selectHolonBySlug(slug: String, reload: Boolean = true) {
        val identity = state.value.availableHolons.firstOrNull { it.slug == slug }
            ?: throw IllegalStateException("Holon '$slug' not found")
        if (state.value.selectedHolon == identity) {
            if (reload) {
                loadLanguages()
            }
            return
        }
        _state.update { it.copy(selectedHolon = identity) }
        stop()
        if (reload) {
            loadLanguages()
        }
    }

    suspend fun setTransport(value: String, reload: Boolean = true) {
        val normalized = normalizedTransportSelection(value)
        require(capabilities.appTransports.contains(normalized)) {
            "Transport \"$normalized\" is not available on this platform"
        }
        if (normalized == state.value.transport) {
            return
        }
        _state.update { it.copy(transport = normalized) }
        stop()
        if (reload) {
            loadLanguages()
        }
    }

    suspend fun setSelectedLanguage(code: String, greetNow: Boolean = true) {
        if (code == state.value.selectedLanguageCode) {
            return
        }
        _state.update { it.copy(selectedLanguageCode = code) }
        if (greetNow) {
            greet()
        }
    }

    suspend fun setUserName(value: String, greetNow: Boolean = true) {
        if (value == state.value.userName) {
            return
        }
        _state.update { it.copy(userName = value) }
        if (greetNow && state.value.selectedLanguageCode.isNotBlank()) {
            greet()
        }
    }

    suspend fun loadLanguages(greetAfterLoad: Boolean = true) {
        val generation = ++loadGeneration
        _state.update {
            it.copy(
                isLoading = true,
                error = null,
                greeting = "",
                availableLanguages = emptyList(),
            )
        }

        try {
            ensureStarted()
        } catch (_: Throwable) {
            if (loadGeneration == generation) {
                _state.update {
                    it.copy(
                        isLoading = false,
                        error = "Failed to load languages: ${it.connectionError ?: "Holon did not become ready"}",
                    )
                }
            }
            return
        }

        if (!state.value.isRunning) {
            if (loadGeneration == generation) {
                _state.update {
                    it.copy(
                        isLoading = false,
                        error = "Failed to load languages: ${it.connectionError ?: "Holon did not become ready"}",
                    )
                }
            }
            return
        }

        val retryDelays = if (state.value.transport == "stdio") {
            listOf(0L, 80L, 180L)
        } else {
            listOf(120L, 300L, 600L)
        }

        retryDelays.forEachIndexed { index, delayMillis ->
            try {
                if (delayMillis > 0) {
                    delay(delayMillis)
                }
                val languages = requireNotNull(connection).listLanguages()
                if (loadGeneration != generation) {
                    return
                }
                val preferredCode = state.value.selectedLanguageCode
                val selectedCode = languages.firstOrNull { it.code == preferredCode }?.code
                    ?: languages.firstOrNull { it.code == "en" }?.code
                    ?: languages.firstOrNull()?.code
                    .orEmpty()
                _state.update {
                    it.copy(
                        availableLanguages = languages,
                        selectedLanguageCode = selectedCode,
                        error = null,
                        isLoading = false,
                    )
                }
                if (greetAfterLoad && selectedCode.isNotBlank()) {
                    greet()
                }
                return
            } catch (error: Throwable) {
                if (index == retryDelays.lastIndex && loadGeneration == generation) {
                    _state.update {
                        it.copy(
                            isLoading = false,
                            error = "Failed to load languages: ${it.connectionError ?: error}",
                        )
                    }
                }
            }
        }
    }

    suspend fun greet(name: String? = null, langCode: String? = null) {
        val resolvedCode = langCode ?: state.value.selectedLanguageCode
        if (resolvedCode.isBlank()) {
            return
        }

        val requestGeneration = ++greetGeneration
        _state.update { it.copy(isGreeting = true) }
        try {
            ensureStarted()
            val response = requireNotNull(connection).sayHello(
                name = name ?: state.value.userName,
                langCode = resolvedCode,
            )
            if (greetGeneration == requestGeneration) {
                _state.update {
                    it.copy(
                        greeting = response,
                        error = null,
                    )
                }
            }
        } catch (error: Throwable) {
            if (greetGeneration == requestGeneration) {
                _state.update {
                    it.copy(error = "Greeting failed: $error")
                }
            }
        } finally {
            if (greetGeneration == requestGeneration) {
                _state.update { it.copy(isGreeting = false) }
            }
        }
    }

    suspend fun ensureStarted() {
        connection?.let { return }
        startFuture?.let {
            it.await()
            connection?.let { return }
        }

        if (state.value.availableHolons.isEmpty()) {
            refreshHolons()
        }

        val holon = state.value.selectedHolon ?: preferredHolon(state.value.availableHolons)
        if (holon == null) {
            _state.update {
                it.copy(
                    isRunning = false,
                    connectionError = "No Gabriel holons found",
                )
            }
            throw IllegalStateException(state.value.connectionError)
        }

        _state.update { it.copy(selectedHolon = holon, connectionError = null) }
        val generation = ++connectionGeneration
        val future = scope.async {
            connect(generation, holon)
        }
        startFuture = future
        try {
            future.await()
        } finally {
            if (startFuture === future) {
                startFuture = null
            }
        }
    }

    private suspend fun connect(generation: Int, holon: GabrielHolonIdentity) {
        try {
            val nextConnection = connector.connect(holon, state.value.transport)
            if (generation != connectionGeneration || disposed) {
                nextConnection.close()
                return
            }
            connection = nextConnection
            _state.update {
                it.copy(
                    isRunning = true,
                    connectionError = null,
                )
            }
        } catch (error: Throwable) {
            if (generation != connectionGeneration || disposed) {
                return
            }
            connection = null
            _state.update {
                it.copy(
                    isRunning = false,
                    connectionError = "Failed to start Gabriel holon: $error",
                )
            }
            throw error
        }
    }

    suspend fun stop() {
        connectionGeneration += 1
        startFuture = null
        val currentConnection = connection
        connection = null
        _state.update { it.copy(isRunning = false) }
        if (currentConnection == null) {
            return
        }
        try {
            currentConnection.close()
        } catch (error: Throwable) {
            _state.update { it.copy(connectionError = "Failed to stop Gabriel holon connection: $error") }
        }
    }

    suspend fun shutdown() {
        disposed = true
        stop()
        scope.cancel()
    }
}

File: examples/hello-world/gabriel-greeting-app-flutter/app/lib/src/controller/greeting_controller.dart
import 'package:holons/gen/holons/v1/coax.pb.dart';
import 'package:holons/gen/holons/v1/manifest.pb.dart';
import 'package:holons/holons.dart' as holons_obs;
import 'package:holons_app/holons_app.dart';

import '../gen/v1/greeting.pb.dart';
import '../model/app_model.dart';
import '../runtime/greeting_holon_connection.dart';

class GreetingController
    extends
        HolonOrchestratorController<
          GabrielHolonIdentity,
          GreetingHolonConnection
        > {
  static const listLanguagesMethod =
      'greeting.v1.GreetingService/ListLanguages';
  static const sayHelloMethod = 'greeting.v1.GreetingService/SayHello';

  GreetingController({
    Holons<GabrielHolonIdentity>? holons,
    @Deprecated('Use holons') HolonCatalog<GabrielHolonIdentity>? catalog,
    required GreetingHolonConnectionFactory connector,
    super.capabilities,
    super.initialTransport,
  }) : assert(holons != null || catalog != null),
       super(
         holons: holons ?? catalog!,
         connector: connector.connect,
         connectionCloser: (connection) => connection.close(),
         slugOf: (holon) => holon.slug,
         buildRunnerOf: (holon) => holon.buildRunner,
         preferredHolon: preferredHolon,
         memberInfoBuilder: _gabrielMemberInfo,
         discoveryErrorPrefix: 'Failed to discover Gabriel holons',
         noHolonsMessage: 'No Gabriel holons found',
         connectionErrorPrefix: 'Failed to start Gabriel holon',
         stopErrorPrefix: 'Failed to stop Gabriel holon connection',
       );

  holons_obs.Logger? _logger;
  holons_obs.Logger? _responseLogger;
  holons_obs.Counter? _sayHelloTotal;
  holons_obs.Histogram? _sayHelloDuration;

  bool _initialized = false;
  int _loadGeneration = 0;
  int _greetGeneration = 0;

  bool isLoading = true;
  bool isGreeting = false;
  String greeting = '';
  String userName = '';
  String selectedLanguageCode = '';
  List<Language> availableLanguages = const <Language>[];

  void attachObservability(holons_obs.Observability observability) {
    _logger = observability.logger('greeting-controller');
    _responseLogger = observability.logger('');
    _sayHelloTotal = observability.counter(
      'gabriel_greeting_say_hello_total',
      help: 'Greeting requests sent from the Flutter app',
      labels: const {'origin': 'app'},
    );
    _sayHelloDuration = observability.histogram(
      'gabriel_greeting_say_hello_duration_seconds',
      help: 'Greeting request duration observed by the Flutter app',
      labels: const {'origin': 'app'},
    );
  }

  String get statusTitle {
    if (isLoading) {
      return 'Starting holon...';
    }
    if (isRunning) {
      return 'Ready';
    }
    return 'Offline';
  }

  @override
  Future<void> initialize() async {
    if (_initialized) {
      return;
    }
    _initialized = true;
    if (userName.isEmpty) {
      userName = 'World';
    }
    notifyListeners();
    await refreshHolons();
    await loadLanguages(greetAfterLoad: false);
    if (selectedLanguageCode.isNotEmpty) {
      await greet();
    }
  }

  Future<void> setSelectedLanguage(String code, {bool greetNow = true}) async {
    if (code == selectedLanguageCode) {
      return;
    }
    selectedLanguageCode = code;
    safeNotify();
    if (greetNow) {
      await greet();
    }
  }

  Future<void> setUserName(String value, {bool greetNow = true}) async {
    if (value == userName) {
      return;
    }
    userName = value;
    safeNotify();
    if (greetNow && selectedLanguageCode.isNotEmpty) {
      await greet();
    }
  }

  Future<void> loadLanguages({bool greetAfterLoad = true}) async {
    final generation = ++_loadGeneration;
    final effectiveTransport = selectedHolon == null
        ? transport
        : effectiveHolonTransport(
            requestedTransport: transport,
            buildRunner: selectedHolon!.buildRunner,
          );
    isLoading = true;
    error = null;
    greeting = '';
    availableLanguages = const <Language>[];
    safeNotify();

    final retryDelays = effectiveTransport == HolonTransportName.stdio.rawValue
        ? const <Duration>[Duration.zero, Duration(milliseconds: 400)]
        : const <Duration>[
            Duration.zero,
            Duration(milliseconds: 200),
            Duration(milliseconds: 800),
          ];

    for (var index = 0; index < retryDelays.length; index += 1) {
      try {
        final delay = retryDelays[index];
        if (delay > Duration.zero) {
          await Future<void>.delayed(delay);
        }
        await ensureStarted();
        final connection = activeConnection;
        if (!isRunning || connection == null) {
          throw StateError(connectionError ?? 'Holon did not become ready');
        }
        final languages = await connection.listLanguages();
        if (_loadGeneration != generation) {
          return;
        }
        availableLanguages = languages;
        final preferredCode = selectedLanguageCode;
        selectedLanguageCode = availableLanguages
            .firstWhere(
              (language) => language.code == preferredCode,
              orElse: () => Language(),
            )
            .code
            .ifEmpty(
              availableLanguages
                  .firstWhere(
                    (language) => language.code == 'en',
                    orElse: () => Language(),
                  )
                  .code
                  .ifEmpty(
                    availableLanguages.isEmpty
                        ? ''
                        : availableLanguages.first.code,
                  ),
            );
        error = null;
        isLoading = false;
        safeNotify();
        if (greetAfterLoad && selectedLanguageCode.isNotEmpty) {
          await greet();
        }
        return;
      } on Object catch (loadError) {
        await dropConnection();
        if (index == retryDelays.length - 1 && _loadGeneration == generation) {
          error =
              'Failed to load languages: ${connectionError ?? loadError.toString()}';
          isLoading = false;
          safeNotify();
        }
      }
    }
  }

  Future<void> greet({String? name, String? langCode}) async {
    final resolvedCode = langCode ?? selectedLanguageCode;
    if (resolvedCode.trim().isEmpty) {
      return;
    }

    final requestGeneration = ++_greetGeneration;
    isGreeting = true;
    safeNotify();
    final startedAt = DateTime.now();
    try {
      await ensureStarted();
      final connection = activeConnection;
      if (connection == null) {
        throw StateError(connectionError ?? 'Holon did not become ready');
      }
      final response = await connection.sayHello(
        name: name ?? userName,
        langCode: resolvedCode,
      );
      if (_greetGeneration != requestGeneration) {
        return;
      }
      greeting = response;
      error = null;
      _sayHelloTotal?.inc();
      _sayHelloDuration?.observe(
        DateTime.now().difference(startedAt).inMicroseconds /
            Duration.microsecondsPerSecond,
      );
      (_responseLogger ?? _logger)?.info(
        'Greeting response received',
        fields: {
          'method': sayHelloMethod,
          'lang': resolvedCode,
          'holon': selectedHolon?.slug ?? '',
        },
      );
    } on Object catch (greetError) {
      if (_greetGeneration != requestGeneration) {
        return;
      }
      error = 'Greeting failed: $greetError';
      _logger?.error(
        'Greeting request failed',
        fields: {
          'method': sayHelloMethod,
          'lang': resolvedCode,
          'holon': selectedHolon?.slug ?? '',
          'error': greetError.toString(),
        },
      );
    } finally {
      if (_greetGeneration == requestGeneration) {
        isGreeting = false;
        safeNotify();
      }
    }
  }

  @override
  Future<void> reloadSelectedHolon() {
    return loadLanguages();
  }

  @override
  Future<Object?> handleMemberTell({
    required String slug,
    required String method,
    Object? payloadJson,
  }) async {
    final decodedPayload = payloadJson ?? const <String, Object?>{};

    switch (method) {
      case listLanguagesMethod:
        await loadLanguages(greetAfterLoad: false);
        return ListLanguagesResponse(
          languages: availableLanguages,
        ).toProto3Json();

      case sayHelloMethod:
        final request = SayHelloRequest()..mergeFromProto3Json(decodedPayload);
        if (request.name.isNotEmpty) {
          await setUserName(request.name, greetNow: false);
        }
        if (request.langCode.isNotEmpty) {
          await setSelectedLanguage(request.langCode, greetNow: false);
        }
        await greet(
          name: request.name.isEmpty ? null : request.name,
          langCode: request.langCode.isEmpty ? null : request.langCode,
        );
        return SayHelloResponse(greeting: greeting).toProto3Json();
    }

    await ensureStarted();
    final connection = activeConnection;
    if (connection == null) {
      throw StateError(connectionError ?? 'Holon did not become ready');
    }
    return connection.tell(method: method, payload: decodedPayload);
  }

  @override
  void didConnectHolon(GabrielHolonIdentity holon, String transportLabel) {
    _logger?.info(
      'Holon connection ready',
      fields: {'holon': holon.slug, 'transport': transportLabel},
    );
  }

  @override
  void didFailHolonConnection(
    GabrielHolonIdentity holon,
    String effectiveTransport,
    Object error,
  ) {
    _logger?.error(
      'Holon connection failed',
      fields: {
        'holon': holon.slug,
        'transport': effectiveTransport,
        'error': error.toString(),
      },
    );
  }
}

MemberInfo _gabrielMemberInfo(
  GabrielHolonIdentity identity,
  MemberState state,
) {
  return MemberInfo(
    slug: identity.slug,
    identity: HolonManifest_Identity(
      familyName: identity.familyName,
      givenName: identity.displayName,
    ),
    state: state,
    isOrganism: false,
  );
}

extension on String {
  String ifEmpty(String fallback) => trim().isEmpty ? fallback : this;
}
